server/src/main/java/org/apache/cloudstack/acl/ProjectRoleManagerImpl.java

```diff
@@ -0,0 +1,308 @@
+        if (projRolePermissionId == null) {
+            return null;
+        }
+        return projRolePermissionsDao.findById(projRolePermissionId);
+    }
+
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_PROJECT_ROLE_PERMISSION_DELETE, eventDescription = "deleting Project Role Permission")
+    public boolean deleteProjectRolePermission(ProjectRolePermission projectRolePermission) {
+        checkAccess(projectRolePermission.getProjectId());
+        return projRolePermissionsDao.remove(projectRolePermission.getId());
+    }
+
+    @Override
+    public List<ProjectRolePermission> findAllProjectRolePermissions(Long projectId, Long projectRoleId) {
+        List<? extends ProjectRolePermission> permissions = projRolePermissionsDao.findAllByRoleIdSorted(projectRoleId, projectId);
+        if (permissions != null) {
+            return new ArrayList<>(permissions);
+        }
+        return Collections.emptyList();
+    }
+
+    @Override
+    public List<ProjectRole> findProjectRolesByName(Long projectId, String roleName) {
+        List<? extends ProjectRole> roles = null;
+        if (StringUtils.isNotBlank(roleName)) {
+            roles = projRoleDao.findByName(roleName, projectId);
+        }
+        return ListUtils.toListOfInterface(roles);
+    }
+
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_PROJECT_ROLE_DELETE, eventDescription = "deleting Project Role")
+    public boolean deleteProjectRole(ProjectRole role, Long projectId) {
+        checkAccess(projectId);
+        if (role == null) {
+            return false;
+        }
+
+        Long roleProjectId = role.getProjectId();
+        if (role.getProjectId() != null && !roleProjectId.equals(projectId)) {
+            throw new PermissionDeniedException("Not authorized to delete the given project role");
+        }
+
+        List<? extends ProjectAccount> users = projAccDao.listUsersOrAccountsByRole(role.getId());
+        if (users != null && users.size() != 0) {
+            throw new PermissionDeniedException("Found users that have the project role in use, cannot delete the Project Role");
+        }
+        return Transaction.execute(new TransactionCallback<Boolean>() {
+            @Override
+            public Boolean doInTransaction(TransactionStatus status) {
+                List<? extends ProjectRolePermission> rolePermissions = projRolePermissionsDao.findAllByRoleIdSorted(role.getId(), projectId);
+                if (rolePermissions != null && !rolePermissions.isEmpty()) {
+                    for (ProjectRolePermission rolePermission : rolePermissions) {
+                        projRolePermissionsDao.remove(rolePermission.getId());
+                    }
+                }
+                if (projRoleDao.remove(role.getId())) {
+                    ProjectRoleVO projRoleVO = projRoleDao.findByIdIncludingRemoved(role.getId());
+                    projRoleVO.setName(null);
```

Comment:
Why is this required after deletion?

Reply:
Only to clear the name pf the role from the DB once its been removed. May not be required, but followed what was done for Roles

Reply:
I'm ok with leaving the update for consistency